src/main/java/gregtech/api/capability/impl/BoilerRecipeLogic.java

```diff
@@ -0,0 +1,282 @@
+import gregtech.common.metatileentities.multi.MetaTileEntityLargeBoiler;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.NonNullList;
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.items.IItemHandlerModifiable;
+
+import java.util.Collections;
+
+import static gregtech.api.capability.GregtechDataCodes.BOILER_HEAT;
+import static gregtech.api.capability.GregtechDataCodes.BOILER_LAST_TICK_STEAM;
+
+// TODO Move to common
+
+public class BoilerRecipeLogic extends AbstractRecipeLogic {
+
+    private static final long STEAM_PER_WATER = 160;
+
+    private int currentHeat;
+    private int lastTickSteamOutput;
+    private int excessWater, excessFuel, excessProjectedEU;
+
+    public BoilerRecipeLogic(MetaTileEntityLargeBoiler tileEntity) {
+        super(tileEntity, null);
+    }
+
+    @Override
+    public void update() {
+        if (!isActive() && currentHeat > 0) {
+            setHeat(currentHeat - 1);
+            setLastTickSteam(0);
+        }
+        super.update();
+    }
+
+    @Override
+    protected void trySearchNewRecipe() {
+        MetaTileEntityLargeBoiler boiler = (MetaTileEntityLargeBoiler) metaTileEntity;
+        if (ConfigHolder.machines.enableMaintenance && boiler.hasMaintenanceMechanics() && boiler.getMaintenanceProblems() > 5) {
+            return;
+        }
+
+        IMultipleTankHandler importFluids = boiler.getImportFluids();
+        boolean didStartRecipe = false;
+
+        // todo can optimize with an override of checkPreviousRecipe() and a check here
+        // (might need to?)
+
+        // TODO Do these after fuel recipes are done differently
+/*
+        for (IFluidTank fluidTank : importFluids.getFluidTanks()) {
+            FluidStack fuelStack = fluidTank.drain(Integer.MAX_VALUE, false);
+            if (fuelStack == null || ModHandler.isWater(fuelStack)) continue;
+
+            FuelRecipe dieselRecipe = RecipeMaps.COMBUSTION_GENERATOR_FUELS.findRecipe(GTValues.V[GTValues.MAX], fuelStack);
+            if (dieselRecipe != null) {
+                // should exit here
+                //int duration = dieselRecipe.getDuration();
+                //Recipe recipe = new Recipe(Collections.emptyList(), Collections.emptyList(), Collections.emptyList(), )
+            }
+
+            FuelRecipe denseFuelRecipe = RecipeMaps.SEMI_FLUID_GENERATOR_FUELS.findRecipe(GTValues.V[GTValues.MAX], fuelStack);
+            if (denseFuelRecipe != null) {
+                // should exit here
+            }
+        }
+*/
+        if (!didStartRecipe) {
+            IItemHandlerModifiable importItems = boiler.getImportItems();
+            for (int i = 0; i < importItems.getSlots(); i++) {
+                ItemStack stack = importItems.getStackInSlot(i);
+                int fuelBurnTime = (int) Math.ceil(ModHandler.getFuelValue(stack));
+                if (fuelBurnTime / 80 > 0) {
```

Comment:
Some comment on why 80 would be nice, even if it is just "This is the chosen divisor"